program/rust/src/tests/test_upd_price.rs

```diff
@@ -0,0 +1,292 @@
+use solana_program::program_error::ProgramError;
+use solana_program::pubkey::Pubkey;
+use std::mem::size_of;
+
+use crate::c_oracle_header::{
+    cmd_upd_price_t,
+    command_t_e_cmd_upd_price,
+    pc_price_t,
+    PC_STATUS_TRADING,
+    PC_STATUS_UNKNOWN,
+    PC_VERSION,
+};
+
+use crate::deserialize::{
+    initialize_pyth_account_checked,
+    load_checked,
+    load_mut,
+};
+use crate::rust_oracle::upd_price;
+use crate::tests::test_utils::{
+    update_clock_slot,
+    AccountSetup,
+};
+use crate::utils::pubkey_assign;
+#[test]
+fn test_upd_price() {
+    let mut instruction_data = [0u8; size_of::<cmd_upd_price_t>()];
+    populate_instruction(&mut instruction_data, Some(42), Some(2), Some(1));
+
+    let program_id = Pubkey::new_unique();
+
+    let mut funding_setup = AccountSetup::new_funding();
+    let funding_account = funding_setup.to_account_info();
+
+    let mut price_setup = AccountSetup::new::<pc_price_t>(&program_id);
+    let mut price_account = price_setup.to_account_info();
+    price_account.is_signer = false;
+    initialize_pyth_account_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+
+    {
+        let mut price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+        price_data.num_ = 1;
+        pubkey_assign(
+            &mut price_data.comp_[0].pub_,
+            &funding_account.key.to_bytes(),
+        );
+    }
+
+    let mut clock_setup = AccountSetup::new_clock();
+    let mut clock_account = clock_setup.to_account_info();
+    clock_account.is_signer = false;
+    clock_account.is_writable = false;
+
+    update_clock_slot(&mut clock_account, 1);
+
+    assert!(upd_price(
+        &program_id,
+        &[
+            funding_account.clone(),
+            price_account.clone(),
+            clock_account.clone()
+        ],
+        &instruction_data
+    )
+    .is_ok());
+
+    {
+        let price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+        assert_eq!(price_data.comp_[0].latest_.price_, 42);
+        assert_eq!(price_data.comp_[0].latest_.conf_, 2);
+        assert_eq!(price_data.comp_[0].latest_.pub_slot_, 1);
+        assert_eq!(price_data.comp_[0].latest_.status_, PC_STATUS_TRADING);
+        assert_eq!(price_data.valid_slot_, 0);
+        assert_eq!(price_data.agg_.pub_slot_, 1);
+        assert_eq!(price_data.agg_.price_, 0);
+        assert_eq!(price_data.agg_.status_, PC_STATUS_UNKNOWN);
+    }
+
+    // add some prices for current slot - get rejected
+    populate_instruction(&mut instruction_data, Some(43), None, None);
```

Comment:
fwiw even here i think this code would be clearer if it required providing every argument. It's less state that the reader has to keep in their head.